pkg/runctl/controller.go

```diff
@@ -166,7 +167,15 @@ func (c *Controller) createRunManager(pipelineRun k8s.PipelineRun, pipelineRunsC
 	tenant := k8s.NewTenantNamespace(c.factory, pipelineRun.GetNamespace())
 	workFactory := tenant.TargetClientFactory()
 	namespaceManager := k8s.NewNamespaceManager(c.factory, runNamespacePrefix, runNamespaceRandomLength)
-	return NewRunManager(workFactory, pipelineRunsConfig, tenant.GetSecretProvider(), namespaceManager)
+	runManagerInstance := NewRunManager(workFactory, pipelineRunsConfig, tenant.GetSecretProvider(), namespaceManager)
+
+	if c.testing != nil && c.testing.runManagerServiceAccountSecretName != "" {
+		rm, ok := runManagerInstance.(*runManager)
+		if ok {
+			rm.testing = &runManagerTesting{getServiceAccountSecretNameStub: func(ctx *runContext) string { return c.testing.runManagerServiceAccountSecretName }}
+		}
+	}
+	return runManagerInstance
```

Comment:
I don't like that there is test setup code mixed into production code.

Better:

- Extract the call to `NewRunManager` into a separate method, which allows to be overridden by a stub in `controllerTesting`.
- In a test inject a stub which returns a `runManager` configured to its needs.

Reply:
Done